### ISSUE TITLE
Phpmyadmin very slow with RDS (MariaDB 10.11) on another region

### ISSUE THREAD

**Issue by rahulthackkar** [phpmyadmin/phpmyadmin#19313]
### Describe the bug
Phpmyadmin is slow in cross AWS Region

### To Reproduce

We are using AWS RDS Instance on ap-southeast-2 (Sydney) and phpmyadmin is hosted on region (us-east-1) as my domain control panel is here on region us-east-1, we have many RDS across regions so difficult to setup PMA on each region.

### Expected behavior

It should not be slow, it is taking a minute or more even while loading few databases also. 
**Though mysql cli is working smooth.**
### Screenshots

N/A

### Server configuration

- Operating system: Ubuntu 22.04
- Web server:Apache2.4.54
- Database version:MariaDB 10.11
- PHP version:8.2
- phpMyAdmin version:5.2.1

### Client configuration

- Browser: Google Chrome
- Operating system:Ubuntu 22.04

### Additional context

Please find attached file for profiling data.
[Uploading profile.txt…]()

Your help will be appreciable.

**Comment by kamil-tekiela** [comment]
Did you try using the master branch 6.0 version?

**Comment by williamdes** [comment]
Also, you may want to add
```php
$cfg['DBG']['sql'] = true;
$cfg['DBG']['sqllog'] = true;
```
And then is it easier to find the slow requests
![image](https://github.com/user-attachments/assets/5fdcdf32-94bb-4964-bfc9-757709c36c25)
But still not 100% easy, but it would help us diagnose.
I share the feeling as I use RDS daily, and it is horribly slow. Probably due to resolving foreigns, but that was improved a lot in 6.0

**Comment by rahulthackkar** [comment]
> Did you try using the master branch 6.0 version?



Not yet, @williamdes Should I try with that version even before releasing TAG?

**Comment by rahulthackkar** [comment]
> Also, you may want to add
> 
> ```
> $cfg['DBG']['sql'] = true;
> $cfg['DBG']['sqllog'] = true;
> ```
> 
> And then is it easier to find the slow requests ![image](https://private-user-images.githubusercontent.com/7784660/375058057-5fdcdf32-94bb-4964-bfc9-757709c36c25.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg1NDY2MDUsIm5iZiI6MTcyODU0NjMwNSwicGF0aCI6Ii83Nzg0NjYwLzM3NTA1ODA1Ny01ZmRjZGYzMi05NGJiLTQ5NjQtYmZjOS03NTc3MDljMzZjMjUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTBUMDc0NTA1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGUyYjBhMDZhNGVmMmRhZGEyMWJkZGQwZGE3YzJmNjczY2MyMzgyYmZlMGZlNTNjMzBkMmMzMDBkYzY5NTIyZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.5C5e6TH_6bByrbhiDOcmoKK3DhR3H4PiWwAq-0GCZCo) But still not 100% easy, but it would help us diagnose. I share the feeling as I use RDS daily, and it is horribly slow. Probably due to resolving foreigns, but that was improved a lot in 6.0

Please find attached, here 'xyz' is altered database name.
I observed that each query is taking minimum time of 0.19xxxx which is typical latency between us-east-1  and ap-southeast-2.

![Screenshot from 2024-10-10 13-13-35](https://github.com/user-attachments/assets/6815ac83-631e-478d-9664-7e1a1de192da)

**Comment by rahulthackkar** [comment]
> Did you try using the master branch 6.0 version?

Even I tried with master branch setup, still facing same issue.

**Comment by williamdes** [comment]
> > Did you try using the master branch 6.0 version?
> 
> Not yet, @williamdes Should I try with that version even before releasing TAG?

Yes

The latest 6.0 version in development [(phpMyAdmin 6.0+snapshot)](https://www.phpmyadmin.net/downloads/#snapshot_6.0+snapshot) can be found on our website. This will be easier than trying to build it from the branch

**Comment by rahulthackkar** [comment]
> On my setup I saved some useless queries by disabling some features
> 
> ```
> $cfg['VersionCheck'] = false;// Save some network latency
> 
> $cfg['NavigationTreeShowFunctions'] = false;
> $cfg['NavigationTreeShowProcedures'] = false;
> $cfg['NavigationTreeShowEvents'] = false;
> 
> $cfg['Servers'][$i]['DisableIS'] = true;// This helps sometimes
> 
> 
> foreach ($cfg['Servers'] as $i => $server) {
>     $cfg['Servers'][$i]['bookmarktable'] = false; // 'pma__bookmark';
>     $cfg['Servers'][$i]['relation'] = false; // 'pma__relation';
>     $cfg['Servers'][$i]['table_info'] = false; // 'pma__table_info';
>     $cfg['Servers'][$i]['table_coords'] = false; // 'pma__table_coords';
>     $cfg['Servers'][$i]['pdf_pages'] = false; // 'pma__pdf_pages';
>     $cfg['Servers'][$i]['column_info'] = false; // 'pma__column_info';
>     $cfg['Servers'][$i]['history'] = false; // 'pma__history';
>     $cfg['Servers'][$i]['table_uiprefs'] = false; // 'pma__table_uiprefs';
>     $cfg['Servers'][$i]['tracking'] = false; // 'pma__tracking';
>     $cfg['Servers'][$i]['userconfig'] = false; // 'pma__userconfig';
>     $cfg['Servers'][$i]['recent'] = false; // 'pma__recent';
>     $cfg['Servers'][$i]['favorite'] = false; // 'pma__favorite';
>     $cfg['Servers'][$i]['users'] = false; // 'pma__users';
>     $cfg['Servers'][$i]['usergroups'] = false; // 'pma__usergroups';
>     $cfg['Servers'][$i]['navigationhiding'] = false; // 'pma__navigationhiding';
>     $cfg['Servers'][$i]['savedsearches'] = false; // 'pma__savedsearches';
>     $cfg['Servers'][$i]['central_columns'] = false; // 'pma__central_columns';
>     $cfg['Servers'][$i]['designer_settings'] = false; // 'pma__designer_settings';
>     $cfg['Servers'][$i]['export_templates'] = false; // 'pma__export_templates';
> }
> ```

Thank you for sharing this , after disabling these features, it is better than before, but still it is not usable for us.

Taking around 15 seconds to load 11 databases including system databases like information_schema, mysql , and when I click on specific database from left navigation, it is taking around 20 seconds to load all tables (~367).


Can you help me with some more actions we can do to make it smoother.

**Comment by kamil-tekiela** [comment]
Can you also take an xDebug profile? https://xdebug.org/docs/profiler

**Comment by Draghmar** [comment]
Just to add my two cents: I have the same behaviour with RDS. I did try 6.0 branch and it looked like a little bit faster but there are bugs there that makes it too annoying to use so I got back to 5.2.1. Changing some of those setting above looks like have some impact on the performance but I will see if I will miss some things they disabled on daily basis. ;)

My theory is that the issue is related to latency and the amount of queries that are run when showing any page in PMA. For example I've seen that the Foreign Relations will generate a lot of additional queries to make links and tooltips. I even tried to optimize that by looking at how it's done in the PMA code and rewrite some parts, but I lack time to focus enough on it. ;)

So to summarise: I don't think there's an easy solution here.

**Comment by rahulthackkar** [comment]
> Can you also take an xDebug profile? https://xdebug.org/docs/profiler

Please find attached profiling information.
[profile.zip](https://github.com/user-attachments/files/17374625/profile.zip)

**Comment by rahulthackkar** [comment]
> Just to add my two cents: I have the same behaviour with RDS. I did try 6.0 branch and it looked like a little bit faster but there are bugs there that makes it too annoying to use so I got back to 5.2.1. Changing some of those setting above looks like have some impact on the performance but I will see if I will miss some things they disabled on daily basis. ;)
> 
> My theory is that the issue is related to latency and the amount of queries that are run when showing any page in PMA. For example I've seen that the Foreign Relations will generate a lot of additional queries to make links and tooltips. I even tried to optimize that by looking at how it's done in the PMA code and rewrite some parts, but I lack time to focus enough on it. ;)
> 
> So to summarise: I don't think there's an easy solution here.

I feel same, We know that it is because of cross region latency, but I am trying to optimise from both sides, from application as well from AWS Infrastructure level also , If I get something from them.

As this architecture is our need and it is difficult to maintain separate PMA instance near to all RDS  in different region.
However I found better performance with MySQL Workbench and Adminer, But we want to stick with PMA only as we are using it from long time and more familiar to us.

Lets see, how much we can improvise this.

**Comment by kamil-tekiela** [comment]
Thank you for sharing the trace. 

Would you be able to show me the result of the MySQL "profiling" for the query:
```
SELECT 1 FROM mysql.user LIMIT 1
```

Similar to this:
![image](https://github.com/user-attachments/assets/ef69dfdd-545d-43b5-a2df-bf8ea086ff49)

**Comment by rahulthackkar** [comment]
Hello @kamil-tekiela 

Here it is.
![RDS-Sydney-mysql-user-phpMyAdmin-5-2-1-10-15-2024_06_00_PM](https://github.com/user-attachments/assets/d295c17a-5f07-4f70-91fc-2fe24964c0d2)

**Comment by kamil-tekiela** [comment]
All of this points to high latency on the network. Most of the time was spent making the queries to the server. There were 50 queries which is what we would expect from PMA (it mostly depends on how many tables you have in the current DB). 

All the major issues that I can see in the trace file have already been fixed on the master branch in the past. In your case, you would benefit from as little roundtrips to the server as possible. William's suggestion to disable all PMA features is the best because they create a lot of extra traffic. 

Can you tell me what is the response time like with a database with a single table and one with about 50 tables?

**Comment by MauricioFauth** [comment]
> I did try 6.0 branch and it looked like a little bit faster but there are bugs there that makes it too annoying to use

@Draghmar Could you please describe what bugs you are experiencing? Opening a new issue here on GitHub would be even better.

**Comment by rahulthackkar** [comment]
> All of this points to high latency on the network. Most of the time was spent making the queries to the server. There were 50 queries which is what we would expect from PMA (it mostly depends on how many tables you have in the current DB).
> 
> All the major issues that I can see in the trace file have already been fixed on the master branch in the past. In your case, you would benefit from as little roundtrips to the server as possible. William's suggestion to disable all PMA features is the best because they create a lot of extra traffic.
> 
> Can you tell me what is the response time like with a database with a single table and one with about 50 tables?

Sorry ,but could not understand this, response time of what action?

**Comment by Draghmar** [comment]
> > I did try 6.0 branch and it looked like a little bit faster but there are bugs there that makes it too annoying to use
> 
> @Draghmar Could you please describe what bugs you are experiencing? Opening a new issue here on GitHub would be even better.

I did open something about getting option to turn off foreign key queries but that was more as a feature. From what I remember there were two major bugs and both were already present at github:
- tree refreshing whenever anything changed, collapsing already opened branches, which would be basically like page refresh,
- issue with user with proper privileges won't being able to manage existing users in 6.0 branch, when every previous one works without issue.

Those two were major enough for me that I needed to get back to 5.2, even though the 6.0 has all those performance optimization you did. :(

**Comment by kamil-tekiela** [comment]
> Sorry ,but could not understand this, response time of what action?

I assume you opened the page on `?route=/database/structure`, i.e. a listing of all tables in a given schema. I wanted to know how fast the page loads with a single table and multiple tables. You can see the response times in "network" tab in the browser console (F12)

**Comment by rahulthackkar** [comment]
> > Sorry ,but could not understand this, response time of what action?
> 
> I assume you opened the page on `?route=/database/structure`, i.e. a listing of all tables in a given schema. I wanted to know how fast the page loads with a single table and multiple tables. You can see the response times in "network" tab in the browser console (F12)

For `?route=/database/structure`
Around 15 Seconds for database with 367 tables and
8 Seconds for database with single table.

**Comment by rahulthackkar** [comment]
> > > Did you try using the master branch 6.0 version?
> > 
> > 
> > Not yet, @williamdes Should I try with that version even before releasing TAG?
> 
> Yes
> 
> The latest 6.0 version in development [(phpMyAdmin 6.0+snapshot)](https://www.phpmyadmin.net/downloads/#snapshot_6.0+snapshot) can be found on our website. This will be easier than trying to build it from the branch

Even this version takes too much time, around a minute to load a database structure with same above disabled features.

Can you please help me with this?

**Comment by rahulthackkar** [comment]
> > > Sorry ,but could not understand this, response time of what action?
> > 
> > 
> > I assume you opened the page on `?route=/database/structure`, i.e. a listing of all tables in a given schema. I wanted to know how fast the page loads with a single table and multiple tables. You can see the response times in "network" tab in the browser console (F12)
> 
> For `?route=/database/structure` Around 15 Seconds for database with 367 tables and 8 Seconds for database with single table.
> 
> I also found that duplicate queries for "SHOW TABLES from `dbname`" (4) queries and "SHOW TABLES from `phpmyadmin`" - (2) queries

Hello @kamil-tekiela ,

Can we do anything to fix this issue?

**Comment by kamil-tekiela** [comment]
I will look into it, but I think this was already fixed on master.

**Comment by liviuconcioiu** [comment]
Probably some sort of query caching for common queries that are used everytime by phpMyAdmin, like show tables, users, etc, enabled in config and also with expiration time will help in this case, of course, at the user's risk. This will avoid some roundtrips.

**Comment by rahulthackkar** [comment]
> I will look into it, but I think this was already fixed on master.

I tried with https://www.phpmyadmin.net/downloads/#snapshot_6.0+snapshot , but it is taking even more time, around a minute.

**Comment by kamil-tekiela** [comment]
> > I will look into it, but I think this was already fixed on master.
> 
> I tried with https://www.phpmyadmin.net/downloads/#snapshot_6.0+snapshot , but it is taking even more time, around a minute.

Can you also provide an xDebug trace for 6.0?

**Comment by rahulthackkar** [comment]
> > > I will look into it, but I think this was already fixed on master.
> > 
> > 
> > I tried with https://www.phpmyadmin.net/downloads/#snapshot_6.0+snapshot , but it is taking even more time, around a minute.
> 
> Can you also provide an xDebug trace for 6.0?

Please find attached. Let me know if any more information required.
 
[cachegrind.out.zip](https://github.com/user-attachments/files/17491310/cachegrind.out.zip)

**Comment by kamil-tekiela** [comment]
Can you change your `MaxExactCount` setting to something lower, say 1000, and see if it changes anything?

**Comment by rahulthackkar** [comment]
> MaxExactCount

No difference at all, too slow around 40 seconds .

Here is my additional config for your reference


``` php
$cfg['EnableAutocompleteForTablesAndColumns'] = false;
$cfg['blowfish_secret'] = 'some-secret';
$cfg['ShowColumnCount'] = false; // Disable column count display
$cfg['ShowStats'] = false; // Defines whether or not to display space usage and statistics about databases and tables
$cfg['MaxExactCount'] = 1000;
$cfg['MaxExactCountViews'] = 0;

$cfg['OBGzip'] = true;'
$cfg['VersionCheck'] = false;// Save some network latency

$cfg['NavigationTreeShowFunctions'] = false;
$cfg['NavigationTreeShowProcedures'] = false;
$cfg['NavigationTreeShowEvents'] = false;

foreach ($cfg['Servers'] as $i => $server) {
    $cfg['Servers'][$i]['bookmarktable'] = false; // 'pma__bookmark';
    $cfg['Servers'][$i]['relation'] = false; // 'pma__relation';
    $cfg['Servers'][$i]['table_info'] = false; // 'pma__table_info';
    $cfg['Servers'][$i]['table_coords'] = false; // 'pma__table_coords';
    $cfg['Servers'][$i]['pdf_pages'] = false; // 'pma__pdf_pages';
    $cfg['Servers'][$i]['column_info'] = false; // 'pma__column_info';
    $cfg['Servers'][$i]['history'] = false; // 'pma__history';
    $cfg['Servers'][$i]['table_uiprefs'] = false; // 'pma__table_uiprefs';
    $cfg['Servers'][$i]['tracking'] = false; // 'pma__tracking';
    $cfg['Servers'][$i]['userconfig'] = false; // 'pma__userconfig';
    $cfg['Servers'][$i]['recent'] = false; // 'pma__recent';
    $cfg['Servers'][$i]['favorite'] = false; // 'pma__favorite';
    $cfg['Servers'][$i]['users'] = false; // 'pma__users';
    $cfg['Servers'][$i]['usergroups'] = false; // 'pma__usergroups';
    $cfg['Servers'][$i]['navigationhiding'] = false; // 'pma__navigationhiding';
    $cfg['Servers'][$i]['savedsearches'] = false; // 'pma__savedsearches';
    $cfg['Servers'][$i]['central_columns'] = false; // 'pma__central_columns';
    $cfg['Servers'][$i]['designer_settings'] = false; // 'pma__designer_settings';
    $cfg['Servers'][$i]['export_templates'] = false; // 'pma__export_templates';
    $cfg['Servers'][$i]['persistent'] = true;
    $cfg['Servers'][$i]['DisableIS'] = true;// This helps sometimes
}
```

**Comment by kamil-tekiela** [comment]
Interesting. I will keep on looking for ways to improve the performance on slow connections, but right now I don't know what to tell you. You can try to bring the `MaxExactCount` down to 1 or maybe lower `MaxTableList` to a much smaller number. PMA fetches a count for every table in the list, and since your connection is so slow, it takes a lot of time to fetch it for so many tables. From the trace, it seems that this is what's taking most of the time. 

Any ideas on how to improve the performance of the source code are welcome.

**Comment by kamil-tekiela** [comment]
What is `$cfg['Servers'][$i]['persistent'] = true;`? I can't find any mention of this in the program.

**Comment by rahulthackkar** [comment]
> What is `$cfg['Servers'][$i]['persistent'] = true;`? I can't find any mention of this in the program.

Sorry, it may be a mis-configuration, we can ignore that.

**Comment by rahulthackkar** [comment]
> Interesting. I will keep on looking for ways to improve the performance on slow connections, but right now I don't know what to tell you. You can try to bring the `MaxExactCount` down to 1 or maybe lower `MaxTableList` to a much smaller number. PMA fetches a count for every table in the list, and since your connection is so slow, it takes a lot of time to fetch it for so many tables. From the trace, it seems that this is what's taking most of the time.
> 
> Any ideas on how to improve the performance of the source code are welcome.

I have tried tuning these values also, But setting MaxTableList to lower (25), page loads in 8-9 seconds now.
But it is difficult to traverse to tables, and also left panel table name search is not working.

Let me know any more information I can provide you to better understand the situation, your help will be appreciable.
Thanks

**Comment by abunch** [comment]
I'm experiencing the same issue with an RDS instance and an EC2 instance running docker both in the same AZ.  Neither of the instances is under any type of load (2-3 users with infrequent use).

**Comment by williamdes** [comment]
> I'm experiencing the same issue with an RDS instance and an EC2 instance running docker both in the same AZ. Neither of the instances is under any type of load (2-3 users with infrequent use).

Are you using MySQL Aurora ?
Can you follow the steps above to provide more debug information ?
